File: src/main/java/puck/gui/svg/SVGFrame.java
package puck.gui.svg;

import puck.graph.DependencyGraph;
import puck.graph.io.PrintingOptions;
import puck.gui.PuckControl;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public class SVGFrame extends JFrame implements StackListener {

    static public class SVGConsole {
        int lines = 10;
        int charPerLine = 50;

        public SVGConsole(){
            panel = new JPanel();
            panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
            panel.add(selection);
            panel.add(textArea);
            textArea.setEditable(false);
        }

        private JPanel panel;
        private JLabel selection = new JLabel();
        private JTextArea textArea = new JTextArea(lines, charPerLine);

        void displaySelection(String node){
            if(node.length()>0)
                selection.setText("Selection : " + node);
            else
                selection.setText("No selection");
        }
//        public void setText(String txt) {
//            textArea.setText(txt);
//        }
        public void appendText(String txt) {
            textArea.append(txt+"\n");
        }
    }

    @Override
    public void update(SVGController svgController) {
        undoButton.setEnabled(svgController.canUndo());
        redoButton.setEnabled(svgController.canRedo());
    }

    private final SVGController controller;
    private JButton undoButton;
    private JButton redoButton;

    private JPanel menu = new JPanel();


    private void addUndoRedoButtonToMenu(){
        undoButton =new JButton(new AbstractAction("Undo"){
            @Override
            public void actionPerformed(ActionEvent e) {
                controller.undo();
            }
        });
        undoButton.setEnabled(false);
        menu.add(undoButton);

        redoButton =new JButton(new AbstractAction("Redo"){
            @Override
            public void actionPerformed(ActionEvent e) {
                controller.redo();
            }
        });
        redoButton.setEnabled(false);
        menu.add(redoButton);

    }

    private File chooseFile(){
        JFileChooser chooser = new JFileChooser();
                /*FileNameExtensionFilter filter = new FileNameExtensionFilter(
                        "JPG & GIF Images", "jpg", "gif");
                chooser.setFileFilter(filter);*/
        int returnVal = chooser.showOpenDialog(SVGFrame.this);
        if(returnVal == JFileChooser.APPROVE_OPTION) {
           return chooser.getSelectedFile();
        }
        return null;
    }

    private void addLoadSaveButton(){
        menu.add(new JButton(new AbstractAction("Save"){
            @Override
            public void actionPerformed(ActionEvent e) {
                File f= chooseFile();
                if(f == null) {
                    console.appendText("no file selected");
                }
                controller.saveRecordOnFile(f);
            }
        }));

        menu.add(new JButton(new AbstractAction("Load"){
            @Override
            public void actionPerformed(ActionEvent e) {
                File f= chooseFile();
                if(f == null) {
                    console.appendText("no file selected");
                }
                controller.loadRecord(f);
            }
        }));

    }

    private void addVisibilityCheckBoxesToMenu(){
        final JCheckBox sigCheckBox = new JCheckBox();
        sigCheckBox.setAction(new AbstractAction("Show signatures") {
            @Override
            public void actionPerformed(ActionEvent e) {
                controller.setSignatureVisible(sigCheckBox.isSelected());
            }
        });
        menu.add(sigCheckBox);

        final JCheckBox idCheckBox = new JCheckBox();
        idCheckBox.setAction(new AbstractAction("Show ids") {
            @Override
            public void actionPerformed(ActionEvent e) {
                controller.setIdVisible(idCheckBox.isSelected());
            }
        });
        menu.add(idCheckBox);
    }

    private final SVGConsole console;

    public SVGFrame(InputStream stream,
                    DependencyGraph g,
                    PrintingOptions opts,
                    PuckControl control) throws IOException {
        this.setLayout(new BorderLayout());
        SVGPanel panel = new SVGPanel(SVGController.documentFromStream(stream));

        console = new SVGConsole();

        controller = SVGController.apply(control, g, opts, panel.canvas, console);
        panel.setController(controller);
        controller.registerAsStackListeners(this);

        addVisibilityCheckBoxesToMenu();

        addUndoRedoButtonToMenu();

        addLoadSaveButton();

        menu.add(console.panel);

        menu.add(new JButton(new AbstractAction("Apply") {
            @Override
            public void actionPerformed(ActionEvent e) {
                controller.applyOnCode();
            }
        }));

        this.add(panel, BorderLayout.CENTER);
        this.add(menu, BorderLayout.SOUTH);


        this.setVisible(true);
        this.setMinimumSize(new Dimension(640, 480));
        this.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
    }
}

File: src/main/scala/puck/gui/svg/SVGFrame.scala
package puck.gui.svg

import java.awt.event.ActionEvent
import java.awt.{Dimension, BorderLayout}
import java.io.{File, InputStream}
import javax.swing._

import puck.graph.DependencyGraph
import puck.graph.io.PrintingOptions
import puck.gui.svg.actions.AddNodeAction
import puck.gui.{PuckConsolePanel, TextAreaLogger, PuckControl}
import puck.javaGraph.nodeKind.Package

import scala.swing.Label

class SVGConsole
( val lines: Int = 10,
  val charPerLine: Int = 50) {

  private val selection: Label = new Label

  val panel0 = new PuckConsolePanel(){
    selection +=: contents
    console.rows = lines
    console.columns = charPerLine
  }

  def panel = panel0.peer
  def console = panel0.console

  private[svg] def displaySelection(node: String) : Unit = {
    if (node.length > 0) selection.text = "Selection : " + node
    else selection.text = "No selection"
  }

  def appendText(txt: String) : Unit = {
    console.append(txt + "\n")
  }
}

class SVGFrame
( stream: InputStream,
  g: DependencyGraph,
  opts: PrintingOptions,
  control: PuckControl
  ) extends JFrame with StackListener {

  def update(svgController: SVGController) : Unit = {
    undoAllButton.setEnabled(svgController.canUndo)
    undoButton.setEnabled(svgController.canUndo)
    redoButton.setEnabled(svgController.canRedo)
  }

  def abstractAction(name:String)
                    (action : ActionEvent => Unit) : AbstractAction =
    new AbstractAction(name){
      def actionPerformed(e: ActionEvent) : Unit = action(e)

    }
  def jbutton(name:String)
              (action : ActionEvent => Unit) : JButton =
   new JButton(abstractAction(name)(action))


  def addButtonToMenu(action : AbstractAction) : Unit = {
    val _ = menu.add(new JButton(action))
  }

  def addButtonToMenu(name:String)
                     (action : ActionEvent => Unit) : Unit =
    addButtonToMenu(abstractAction(name)(action))



  private val consolePanel: SVGConsole = new SVGConsole()
  setLayout(new BorderLayout)
  val panel : SVGPanel = new SVGPanel(SVGController.documentFromStream(stream))
  setVisible(true)

  val consoleLogger = new TextAreaLogger(consolePanel.console, g.logger.askPrint)
  private val controller: SVGController = SVGController(control, g.withLogger(consoleLogger), opts, panel.canvas, consolePanel)
  panel.setController(controller)
  private val menu: JPanel = new JPanel()
  controller.registerAsStackListeners(this)

  addVisibilityCheckBoxesToMenu()


  private val undoAllButton = new JButton(abstractAction("Undo all") {
    _ => controller.undoAll()
  })
  undoAllButton.setEnabled(false)
  menu.add(undoAllButton)

  private val undoButton = new JButton(abstractAction("Undo") {
    _ => controller.undo()
  })
  undoButton.setEnabled(false)
  menu.add(undoButton)
  private val redoButton = new JButton(abstractAction("Redo"){
    _ => controller.redo()
  })
  redoButton.setEnabled(false)
  menu.add(redoButton)

  addLoadSaveButton()

  menu.add(consolePanel.panel)

  private def addHboxButtons() : Unit = {
    val hbox = new JPanel()
    hbox.setLayout(new BoxLayout(hbox, BoxLayout.Y_AXIS))
    menu add hbox

    hbox add jbutton("Apply") {
      _ => controller.applyOnCode()
    }

    hbox add new JButton(
      new AddNodeAction(controller.graph.root, controller, Package))


    hbox add jbutton("Show top level packages") {
        _ => controller.expand(controller.graph.rootId)
    }

    val _ = hbox add jbutton("Hide type relationship") {
      _ => controller.setSelectedEdgeForTypePrinting(None)
    }
  }

  addHboxButtons()

  this.add(panel, BorderLayout.CENTER)
  this.add(menu, BorderLayout.SOUTH)


  this.setVisible(true)
  this.setMinimumSize(new Dimension(640, 480))
  this.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE)



  private def chooseFile(chooserMode : (JFileChooser, java.awt.Component) => Int) : Option[File] = {
    val chooser = new JFileChooser()
    chooser.setCurrentDirectory(controller.genController.filesHandler.workingDirectory)
    val returnVal: Int = chooserMode(chooser, SVGFrame.this)
    if (returnVal == JFileChooser.APPROVE_OPTION) {
      Some(chooser.getSelectedFile)
    }
    else None
  }

  private def openFile() : Option[File] =
    chooseFile( (chooser, component) => chooser.showOpenDialog(component) )
  private def saveFile() : Option[File] =
    chooseFile( (chooser, component) => chooser.showSaveDialog(component) )



  private def addLoadSaveButton() : Unit = {
    addButtonToMenu("Save") {
      _ =>
        saveFile() match {
          case None => consolePanel.appendText("no file selected")
          case Some(f) =>  controller.saveRecordOnFile(f)
        }
      }
    addButtonToMenu("Load") {
      _ =>
        openFile() match {
          case None => consolePanel.appendText("no file selected")
          case Some(f) => controller.loadRecord(f)
        }
    }
  }

  private def addVisibilityCheckBoxesToMenu() : Unit = {
    val hbox = new JPanel()
    hbox.setLayout(new BoxLayout(hbox, BoxLayout.Y_AXIS))
    menu add hbox

    val sigCheckBox: JCheckBox = new JCheckBox
    sigCheckBox.setAction(new AbstractAction("Show signatures") {
      def actionPerformed(e: ActionEvent) : Unit = {
        controller.setSignatureVisible(sigCheckBox.isSelected)
      }
    })
    hbox add sigCheckBox
    val idCheckBox: JCheckBox = new JCheckBox
    idCheckBox.setAction(abstractAction("Show ids"){
     _ => controller.setIdVisible(idCheckBox.isSelected)
      })

    hbox add idCheckBox
    val vEdgesCheckBox: JCheckBox = new JCheckBox
    vEdgesCheckBox.setAction(abstractAction("Show Virtual Edges"){
      _ =>
      controller.setVirtualEdgesVisible(vEdgesCheckBox.isSelected)
      })
    hbox add vEdgesCheckBox
    val redOnlyCheckBox: JCheckBox = new JCheckBox
    redOnlyCheckBox.setAction(abstractAction("Show RedOnly"){
      _ =>
      controller.setRedEdgesOnly(redOnlyCheckBox.isSelected)
      })
    hbox add redOnlyCheckBox
    ()
  }


}
